frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -885,6 +885,103 @@ class TypedFnSignature {
   /// \endcond DO_NOT_DOCUMENT
 };
 
+// Container for resolution candidates and (if applicable) their correponding
+// forwarding-to types.
+struct CandidatesAndForwardingInfo {
+  private:
+  std::vector<const TypedFnSignature*> candidates;
+  std::vector<types::QualifiedType> forwardingInfo;
+
+  public:
+   // Add a candidate and optional forwarding info.
+   void addCandidate(const TypedFnSignature* candidate,
+                     const types::QualifiedType* forwardingTo = nullptr) {
+     candidates.push_back(candidate);
+     if (forwardingTo) {
+       forwardingInfo.push_back(*forwardingTo);
+     }
+   }
+
+   // Compute and fill in forwarding info for a range of newly-added candidates.
+   void helpComputeForwardingTo(const CallInfo& fci, size_t start) {
+     types::QualifiedType forwardingReceiverActualType = fci.calledType();
+     forwardingInfo.resize(start);
+     for (size_t i = start; i < candidates.size(); i++) {
+       forwardingInfo.push_back(forwardingReceiverActualType);
+     }
+   }
+
+   // Move the contents of another container into this one, clearing out the
+   // other.
+   void takeFromOther(CandidatesAndForwardingInfo& other) {
+     candidates.insert(candidates.end(),
+                       std::make_move_iterator(other.candidates.begin()),
+                       std::make_move_iterator(other.candidates.end()));
+     forwardingInfo.insert(
+         forwardingInfo.end(),
+         std::make_move_iterator(other.forwardingInfo.begin()),
+         std::make_move_iterator(other.forwardingInfo.end()));
+     other.candidates.clear();
+     other.forwardingInfo.clear();
+   }
+
+   // Get the candidate at the provided index with no bounds checking.
+   const TypedFnSignature* get(size_t i) const { return candidates[i]; }
+
+   // Get the forwarding info at the provided index with no bounds checking.
+   types::QualifiedType getForwardingInfo(size_t i) const { return forwardingInfo[i]; }
```

Comment:
Consider returning by `const types::QualifiedType&` to avoid unnecessary copies.

Reply:
Also, since `forwardingInfo` can be shorter than `candidates`, are you sure you want to forego bounds checking here? Perhaps you need to return `QualifiedType()` if `i < candidates.size()` but `i >= forwardingInfo.size()`?

Reply:
I believe `forwardingInfo` should only be shorter than `candidates` if 1) it's unused, or 2) some new candidates were just inserted and we're just about to `helpComputeForwardingTo` to them. Since that relies on things being called correctly I would like to enforce it with an assertion. What would you think about just asserting `forwardingInfo.size() == candidates.size()` in `getForwardingInfo`, to prevent retrieving anything from a `CandidatesAndForwardingInfo` that's been used incorrectly?